File: 1319-number-of-operations-to-make-network-connected/1319-number-of-operations-to-make-network-connected.py
class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        
        if len(connections) < n - 1:
            return - 1
        
        graph = defaultdict(list)
        for v1, v2 in connections:
            graph[v1].append(v2)
            graph[v2].append(v1)
        visited = set()
        def dfs(node):
            visited.add(node)    
            for nei in graph[node]:
                if nei not in visited:
                    visited.add(node)
                    dfs(nei)
        count = 0
        for i in range(n):
            if i not in visited:
                count+=1
                dfs(i)
        return count - 1

File: 1319-number-of-operations-to-make-network-connected/1319-number-of-operations-to-make-network-connected.cpp
class Solution {
public:
    
    set <int> st;
    
    void dfs(set <int>& st, unordered_map <int, vector<int>>& graph, int node)
    {
        st.insert(node);
        for(auto child : graph[node])
        {
            if(st.find(child) == st.end())
            {
                st.insert(child);
                dfs(st, graph, child);
            }
        }
    }
    
    int makeConnected(int n, vector<vector<int>>& connections) 
    {
        if(connections.size() < n - 1)
            return -1;
        unordered_map <int, vector <int>> graph;
        for(auto v : connections)
        {
            graph[v[0]].push_back(v[1]);
            graph[v[1]].push_back(v[0]);
        }
        int count = 0;
        
        for(int i = 0; i < n; i++)
        {
            if(st.find(i)==st.end())
            {
                dfs(st, graph, i);
                count++;
            }
        }
        
        return count - 1;
    }
};